### DISCUSSION TITLE
ClickHouse - Disk Space Discrepancy Issue

### DISCUSSION THREAD

**Discussion by karthik-thiyagarajan** [ClickHouse/ClickHouse#52528] [Q&A]
We are using "clickhouse/clickhouse:23.3" and we have configured this with 3 shards and 3 replicas. We have used 300 GB disk space for now. While I could see that only 9 to 10 GB is occupied for actual "ClickHouse" Data as per the dashboard, and it appears one of the pods is having almost 80% occupied on the total disk space occupied, while the other pods are still between 5% to 70% only. Can someone help me understand what could be the reason to have heavy usage of disk space in only one pod ?
Any debugging steps would be much useful to identify the root cause.

**Comment by den-crane** [comment]
use these https://kb.altinity.com/altinity-kb-useful-queries/altinity-kb-database-size-table-column-size/
to get sizes of table.
Could be a system database.

**Comment by karthik-thiyagarajan** [comment]
Thanks. I ran these queries and the sizes provided by the first gives me pretty much the right size of the data (~8 gb). However, my total volume occupied is around 200 gb which doesnt make sense and that too it happens in one of those pods. Any idea, how do i figure the extraneous data to identify the root cause ?

**Reply by karthik-thiyagarajan** [reply]
System database just occupies only 2gb, and my data occupies approximately 6 gb.

**Comment by karthik-thiyagarajan** [comment]
Can someone help on this ? System database just occupies only 2gb, and my data occupies approximately 6 gb. However as per the grafana dashboard, my total volume occupied is 200+ GB which i dont know how and how to get rid of that or clean it up.

**Comment by KyoUK4n** [comment]
I am encountering the same issue.
There are 2 shards and 2 replicas in my cluster. Replica-0 in shard-0 is using 830GiB, and replica-1 in shard-1 is using 890GiB. The situation is similar on shard-1 and its replicas.
I've checked the system.parts and system.*_log tables, and they don't occupy too much disk space.
I've also examined other tables that use the ReplicatedMergeTree engine, and there is no discrepancy.
It's confusing. Is there any "dirty data" on the disk? How can I clean it up?

**Reply by den-crane** [reply]
Any evidences that Clickhouse uses this disk space? Check shadow (backup) folder `/var/lib/clickhouse/shadow/`

**Reply by den-crane** [reply]
It could be Distributed tables, check `system.distribution_queue`

**Comment by sky9611** [comment]
Met same issue as well.
CH version: 23.2.4
I did a basic comparison between `system.parts`, `system.disks` and `system.detached_parts` :
``` sql
SELECT
    name,
    formatReadableSize(total_space - unreserved_space) AS reserved_space,
    formatReadableSize(table_size) AS table_parts_size,
    formatReadableSize(detached_part_size) AS detached_size,
    formatReadableSize(((total_space - unreserved_space) - table_size) - detached_part_size) AS extra_space
FROM system.disks AS a
INNER JOIN
(
    SELECT
        disk_name,
        sum(bytes) AS table_size
    FROM system.parts
    GROUP BY disk_name
) AS b ON a.name = b.disk_name
INNER JOIN
(
    SELECT
        disk,
        sum(bytes_on_disk) AS detached_part_size
    FROM system.detached_parts
    GROUP BY disk
) AS c ON b.disk_name = c.disk
```

Result:
|name|reserved_space|table_parts_size|detached_size|extra_space|
|---|---|---|---|---|
| default | 4.21 TiB       | 4.18 TiB         | 18.57 GiB     | 11.67 GiB   |
| disk1   | 3.82 TiB       | 2.53 TiB         | 253.82 GiB    | 1.04 TiB    |
| disk2   | 3.77 TiB       | 1.93 TiB         | 50.98 GiB     | 1.79 TiB    |
| disk3   | 3.58 TiB       | 3.48 TiB         | 5.50 GiB      | 97.38 GiB   |
| disk4   | 4.32 TiB       | 4.31 TiB         | 11.17 GiB     | 1.28 GiB    |

I also checked `system.distribution_queue` and `/var/lib/clickhouse/shadow` path, but there is no records.

**Comment by corradofiore** [comment]
We just encountered a similar situation.  Our cluster had an issue with replication (no space left on the disk of one of the nodes) and the replication recovery process left lots of detached (= old) files on the file system.  This is not a bug, but rather the default behaviour:  https://clickhouse.com/docs/en/operations/settings/merge-tree-settings#detach_old_local_parts_when_cloning_replica

You can see whether you are in this situation by running a query such as:

```sql
SELECT *
FROM system.detached_parts
ORDER BY bytes_on_disk DESC
LIMIT 10
```

The resulting table will give you a hint of the recoverable disk space.

If the reason why the files were detached is `clone`, you can safely delete them with `rm -rf` (bear in mind, `path` refers to a directory, not to a file).  Reference: https://kb.altinity.com/altinity-kb-useful-queries/detached-parts/

Depending on your requirements, you might also want to set `detach_old_local_parts_when_cloning_replica` to `false`, so that no old files will be left hanging around in the event of a replication recovery.

HTH
